File: src/AdventOfCode/Puzzles/Y2022/Day05.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2022;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2022/day/5</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2022, 05, "Supply Stacks", RequiresData = true)]
public sealed class Day05 : Puzzle
{
    /// <summary>
    /// Gets a string containing the IDs of the crates at
    /// the top of each stack using a CraneMover 9000.
    /// </summary>
    public string TopCratesOfStacks9000 { get; private set; } = string.Empty;

    /// <summary>
    /// Gets a string containing the IDs of the crates at
    /// the top of each stack using a CraneMover 9001.
    /// </summary>
    public string TopCratesOfStacks9001 { get; private set; } = string.Empty;

    /// <summary>
    /// Rearranges the crates specified by the puzzle input.
    /// </summary>
    /// <param name="instructions">The instructions of how to rearrange the crates.</param>
    /// <param name="canMoveMultipleCrates">Whether the crane is capable of moving multiple crates at once.</param>
    /// <returns>
    /// A string containing the IDs of the crates at the top of each stack after rearrangement.
    /// </returns>
    public static string RearrangeCrates(IList<string> instructions, bool canMoveMultipleCrates)
    {
        int count = GetStackCount(instructions);
        var stacks = GetStacks(instructions, count);
        var steps = GetSteps(instructions);

        Rearrange(steps, stacks, canMoveMultipleCrates);

        return GetTopCrates(stacks);

        static int GetStackCount(IList<string> instructions)
        {
            for (int i = 0; i < instructions.Count; i++)
            {
                if (string.IsNullOrEmpty(instructions[i]))
                {
                    string trimmed = instructions[i - 1].Trim();
                    return Parse<int>(trimmed[trimmed.LastIndexOf(' ')..]);
                }
            }

            return -1;
        }

        static List<Stack<char>> GetStacks(IList<string> instructions, int count)
        {
            var stacks = new List<Stack<char>>(count);

            for (int i = 0; i < count; i++)
            {
                stacks.Add(new Stack<char>());
            }

            foreach (string line in instructions.Take(count).Reverse())
            {
                for (int i = 1, j = 0; i < line.Length; i += 4, j++)
                {
                    char container = line[i];

                    if (container is ' ')
                    {
                        continue;
                    }

                    stacks[j].Push(container);
                }
            }

            return stacks;
        }

        static List<(int Count, int From, int To)> GetSteps(IList<string> instructions)
        {
            var result = new List<(int Count, int From, int To)>();

            foreach (string instruction in instructions.SkipWhile((p) => !string.IsNullOrEmpty(p)).Skip(1))
            {
                string[] split = instruction.Split(' ');
                result.Add((Parse<int>(split[1]), Parse<int>(split[3]), Parse<int>(split[5])));
            }

            return result;
        }

        static string GetTopCrates(List<Stack<char>> stacks)
        {
            var result = new StringBuilder(stacks.Count);

            foreach (var stack in stacks)
            {
                result.Append(stack.Pop());
            }

            return result.ToString();
        }

        static void Rearrange(
            List<(int Count, int From, int To)> steps,
            List<Stack<char>> stacks,
            bool canMoveMultipleCrates)
        {
            Stack<char>? swap = null;

            foreach ((int count, int from, int to) in steps)
            {
                var source = stacks[from - 1];
                var destination = stacks[to - 1];

                if (canMoveMultipleCrates && count > 1)
                {
                    swap ??= new Stack<char>(count);

                    for (int i = 0; i < count; i++)
                    {
                        swap.Push(source.Pop());
                    }

                    for (int i = 0; i < count; i++)
                    {
                        destination.Push(swap.Pop());
                    }
                }
                else
                {
                    for (int i = 0; i < count; i++)
                    {
                        destination.Push(source.Pop());
                    }
                }
            }
        }
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        var instructions = await ReadResourceAsLinesAsync();

        TopCratesOfStacks9000 = RearrangeCrates(instructions, canMoveMultipleCrates: false);
        TopCratesOfStacks9001 = RearrangeCrates(instructions, canMoveMultipleCrates: true);

        if (Verbose)
        {
            Logger.WriteLine("The crates on the top of each stack with the CraneMover 9000 are: {0}.", TopCratesOfStacks9000);
            Logger.WriteLine("The crates on the top of each stack with the CraneMover 9001 are: {0}.", TopCratesOfStacks9001);
        }

        return PuzzleResult.Create(TopCratesOfStacks9000, TopCratesOfStacks9001);
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2022/Day05.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { from } from 'linq-to-typescript';
import { Puzzle } from '../Puzzle';
import { Puzzle2022 } from './Puzzle2022';

export class Day05 extends Puzzle2022 {
    topCratesOfStacks9000: string;
    topCratesOfStacks9001: string;

    override get name() {
        return 'Supply Stacks';
    }

    override get day() {
        return 5;
    }

    protected override get requiresData() {
        return true;
    }

    static rearrangeCrates(instructions: string[], canMoveMultipleCrates: boolean) {
        const getStackCount = (instructions: string[]) => {
            for (let i = 0; i < instructions.length; i++) {
                if (instructions[i].length === 0) {
                    const trimmed = instructions[i - 1].trim();
                    return Puzzle.parse(trimmed[trimmed.length - 1]);
                }
            }

            return -1;
        };

        type Stack = Array<string>;
        type Step = [count: number, from: number, to: number];

        const getSteps = (instructions: string[]) => {
            const result: Step[] = [];

            const moves: string[] = from(instructions)
                .skipWhile((p: string) => p.length > 0)
                .skip(1)
                .toArray();

            for (const move of moves) {
                const split = move.split(' ');
                result.push([Puzzle.parse(split[1]), Puzzle.parse(split[3]), Puzzle.parse(split[5])]);
            }

            return result;
        };

        const getStacks = (instructions: string[], count: number) => {
            const stacks: Stack[] = [];

            for (let i = 0; i < count; i++) {
                stacks.push([]);
            }

            const lines: string[] = from(instructions).take(count).reverse().toArray();

            for (const line of lines) {
                for (let i = 1, j = 0; i < line.length; i += 4, j++) {
                    const container = line[i];

                    if (container === ' ') {
                        continue;
                    }

                    stacks[j].push(container);
                }
            }

            return stacks;
        };

        const rearrange = (steps: Step[], stacks: Stack[], canMoveMultipleCrates: boolean) => {
            let swap: string[] = [];

            for (const [count, from, to] of steps) {
                const source = stacks[from - 1];
                const destination = stacks[to - 1];

                if (canMoveMultipleCrates && count > 1) {
                    for (let i = 0; i < count; i++) {
                        swap.push(source.pop());
                    }

                    for (let i = 0; i < count; i++) {
                        destination.push(swap.pop());
                    }
                } else {
                    for (let i = 0; i < count; i++) {
                        destination.push(source.pop());
                    }
                }
            }
        };

        const getTopCrates = (stacks: Stack[]) => {
            let result = '';

            for (const stack of stacks) {
                result += stack.pop();
            }

            return result;
        };

        const count = getStackCount(instructions);
        const stacks = getStacks(instructions, count);
        const steps = getSteps(instructions);

        rearrange(steps, stacks, canMoveMultipleCrates);

        return getTopCrates(stacks);
    }

    override solveCore(_: string[]) {
        const moves = this.readResourceAsLines();

        this.topCratesOfStacks9000 = Day05.rearrangeCrates(moves, false);
        this.topCratesOfStacks9001 = Day05.rearrangeCrates(moves, true);

        console.info(`The crates on the top of each stack with the CraneMover 9000 are: ${this.topCratesOfStacks9000}.`);
        console.info(`The crates on the top of each stack with the CraneMover 9001 are: ${this.topCratesOfStacks9001}.`);

        return this.createResult([this.topCratesOfStacks9000, this.topCratesOfStacks9001]);
    }
}
